src/Orchard.Web/Core/Navigation/Services/DefaultMenuProvider.cs

```diff
@@ -27,7 +28,23 @@ public class DefaultMenuProvider : IMenuProvider {
                     .Where(x => x.MenuId == menu.Id)
                     .List();
             }
-            var menuParts = _menuPartsMemory[menu.Id];
+            var menuParts = _menuPartsMemory[menu.Id].ToList<MenuPart>();
+
+            //List of hidden items
+            var menuPartsHidden = _contentManager
```

Comment:
The menus are already loaded from database, why do another query? You can just filter in Linq.

Reply:
I tried but the code didn't return the list I expected, this is the committed code I added in my version

            //An attempt to optimize the code above but unsuccessful
            //var menuParts = _contentManager
            //    .Query<MenuPart, MenuPartRecord>()
            //    .Where(x => x.MenuId == menu.Id && x.VisibleAtFrontEnd && !_contentManager
            //    .Query<MenuPart, MenuPartRecord>()
            //    .Where(x => x.MenuId == menu.Id && !x.VisibleAtFrontEnd)
            //    .List().Where(w => x.MenuPosition.StartsWith(w.MenuPosition)).Any())
            //    .List().ToList();